flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/read/HiveContinuousMonitoringFunction.java

```diff
@@ -0,0 +1,280 @@
+		} else {
+			LOG.info("No state to restore for the {}.", getClass().getSimpleName());
+			this.currentReadTime = this.startupTimestampMillis;
+		}
+	}
+
+	@Override
+	public void run(SourceContext<TimestampedHiveInputSplit> context) throws Exception {
+		checkpointLock = context.getCheckpointLock();
+		while (isRunning) {
+			synchronized (checkpointLock) {
+				monitorAndForwardSplits(context);
+			}
+			Thread.sleep(interval);
+		}
+	}
+
+	private void monitorAndForwardSplits(
+			SourceContext<TimestampedHiveInputSplit> context) throws IOException, TException {
+		assert (Thread.holdsLock(checkpointLock));
+
+		List<Partition> partitions = client.listPartitionsByFilter(
+				tablePath.getDatabaseName(),
+				tablePath.getObjectName(),
+				strategy.generateFetchFilter(partitionKeys, currentReadTime),
+				(short) -1);
+
+		if (partitions.isEmpty()) {
+			return;
+		}
+
+		long maxTime = Long.MIN_VALUE;
+		for (Partition partition : partitions) {
+			List<String> partSpec = partition.getValues();
+			if (!this.distinctPartitions.contains(partSpec)) {
+				this.distinctPartitions.add(partition.getValues());
+				long time = this.strategy.extractPartTime(partitionKeys, partition.getValues());
+				if (time > maxTime) {
+					maxTime = time;
+				}
+				HiveTableInputSplit[] splits = HiveTableInputFormat.createInputSplits(
+						this.readerParallelism,
+						Collections.singletonList(toHiveTablePartition(partition)),
+						this.conf.conf());
+				for (HiveTableInputSplit split : splits) {
+					context.collect(new TimestampedHiveInputSplit(time, split));
+				}
+			}
+		}
+		this.currentReadTime = maxTime;
+
+		this.distinctPartitions.removeIf(partSpec -> this.strategy.canExpireForDistinct(
+				this.strategy.extractPartTime(partitionKeys, partSpec),
+				this.currentReadTime));
+	}
+
+	private HiveTablePartition toHiveTablePartition(Partition p) {
+		return HiveTableSource.toHiveTablePartition(
+				partitionKeys, fieldNames, fieldTypes, hiveShim, tableProps, defaultPartitionName, p);
+	}
+
+	@Override
+	public void snapshotState(FunctionSnapshotContext context) throws Exception {
+		Preconditions.checkState(this.currReadTimeState != null,
+				"The " + getClass().getSimpleName() + " state has not been properly initialized.");
+
+		this.currReadTimeState.clear();
+		this.currReadTimeState.add(this.currentReadTime);
+
+		this.distinctPartsState.clear();
+		this.distinctPartsState.add(new ArrayList<>(this.distinctPartitions));
+
+		if (LOG.isDebugEnabled()) {
+			LOG.debug("{} checkpointed {}.", getClass().getSimpleName(), currentReadTime);
+		}
+	}
+
+	@Override
+	public void close() throws Exception {
+		super.close();
+
+		if (checkpointLock != null) {
+			synchronized (checkpointLock) {
+				currentReadTime = Long.MAX_VALUE;
+				isRunning = false;
+			}
+		}
+	}
+
+	@Override
+	public void cancel() {
+		if (checkpointLock != null) {
+			// this is to cover the case where cancel() is called before the run()
```

Comment:
The `checkpointLock ` is assigned in `run()`. So if cancel is called before run, the `checkpointLock ` should always be null?

Reply:
this is from another thread.